### ISSUE TITLE
DeepSSM in Studio Phase 2

### ISSUE THREAD

**Issue by jadie1** [SCIInstitute/ShapeWorks#1887]
Phase 2 is adding training optimization and image grooming to Studio for DeepSSM. 
Currently to run DeepSSM in Studio you load groomed images and particle pairs. Ideally we'd like to start with ungroomed images and shapes and create the training and testing data. This involves:

- Optimizing training particles
- Optimizing testing particles using fixed domains
- Grooming images
   - Resampling
   - Applying transforms (shape grooming alignment and procrustes if used in training optimization)
   - Cropping (using bounding box that fits world particles plus some padding)

For more detail see the [DeepSSM use case](http://sciinstitute.github.io/ShapeWorks/latest/use-cases/deep-learning-based/deep-ssm-femur.html), but note in Phase 2 we will skip step 7 and prep the testing images in the same way as the training. In Phase 3 we will do true inference and groom testing images using image-to-image registration.

**Comment by zahidemon** [comment]
@akenmorris, I have been talking to @jadie1 about this issue. We already have the Python codes for this phase 2. We just need to figure out how to incorporate them into the studio. We can discuss this more next week in the dev meeting and a separate meeting if you'd like to discuss this earlier than that.

**Comment by jadie1** [comment]
Based on today's discussion, here is a tentative plan:

Define split before opening studio (use case steps 1 and 2)
1. Load training data into studio - data tab
2. Groom training meshes - groom tab (use case step 3)
3. Optimize training particles - optimize tab (use case step 4) 
4. Analyze training particles - analyze tab
5. Load validation data - fixed domain tab 
6. Groom validation meshes using training ref- fixed domain tab (use case step 8)
7. Optimize validation particles using fixed domains - fixed domain tab
8. Train and validation image grooming step 1: apply mesh alignment transforms to images (reflection, alignment, Procrustes)  - image grooming tab or deepssm tab (use case step 5)
9.  Train and validation image grooming step 2: find bounding box using train meshes +transforms, use it to crop all images - image grooming tab or deepssm tab (use case step 5) 
10. Run data augmentation using training images and world particles - deepssm tab (use case step 6)
11. Create data loaders - deepssm tab (use case step 9)
12. Train DeepSSM - deepssm tab (use case step 10)
13. Predict validation particles and analyze accuracy (use case step 11)

Phase 3 will involve adding a test set and incorporating use case step 7 (groom validation and test images using image registration) and use case step 12 ( Predict test particles, true inference). In phase 3, we could also add the option to define the split in the studio data tab. 

The things that we need to implement for phase 2 are:
1.  Fixed domain tab for steps 5, 6, and 7
2.  Image grooming tab (could be part of deepssm tab) for steps 8 and 9

@akenmorris and @zahidemon Do you agree with this rough plan?

**Comment by zahidemon** [comment]
Thanks for summarizing the whole thing, @jadie1. Looks like a solid plan. I think the starting point would be the fixed domain tab.  Can we set up a meeting next week to discuss the fixed domain tab, @akenmorris and @jadie1?

**Comment by akenmorris** [comment]
> If we are not fixing the covariance matrix, but rather just not updating the training particle positions, then this is a separate issue that we should talk about.

Ok, yeah, the fixed domains functionality does not currently work like this.  The effect of setting a domain fixed is primarily that the particles just aren't allowed to move.  All shapes in the system contribute to the covariance matrix.

**Comment by akenmorris** [comment]
Done #2181
